ipld/amt/src/iter.rs

```diff
@@ -0,0 +1,559 @@
+    ///
+    /// Basic usage:
+    ///
+    /// ```rust
+    /// use fvm_ipld_amt::Amt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    ///
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// let mut amt = Amt::new(store);
+    /// amt.set(2, "two".to_owned());
+    /// amt.set(1, "one".to_owned());
+    ///
+    /// let vals = amt.values().collect::<Vec<_>>();
+    /// assert_eq!(vals, ["one", "two"]);
+    ///
+    /// # anyhow::Ok(())
+    /// ```
+    pub fn values(&self) -> impl Iterator<Item = &V> {
+        self.iter().map(|res| {
+            res.map(|(_, v)| v)
+                .expect("Failed to generate iterator from AMT values")
+        })
+    }
+
+    /// Generates an iterator containing the keys of the AMT, in sorted order.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```rust
+    /// use fvm_ipld_amt::Amt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    ///
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// let mut amt = Amt::new(store);
+    /// amt.set(2, "two".to_owned());
+    /// amt.set(1, "one".to_owned());
+    ///
+    /// let keys = amt.keys().collect::<Vec<_>>();
+    /// assert_eq!(keys, [1, 2]);
+    ///
+    /// # anyhow::Ok(())
+    /// ```
+    pub fn keys(&self) -> impl Iterator<Item = usize> + '_ {
+        self.iter().map(|res| {
+            res.map(|(k, _)| k)
+                .expect("Failed to generate iterator from AMT keys")
+        })
+    }
+
+    /// Iterate over the AMT starting at the given key. This can be used to implement "ranged"
+    /// iteration:
+    ///
+    /// ```rust
+    /// use fvm_ipld_amt::Amt;
+    /// use fvm_ipld_blockstore::MemoryBlockstore;
+    ///
+    /// let store = MemoryBlockstore::default();
+    ///
+    /// // Create an AMT with 5 keys.
+    /// let mut amt = Amt::new(store);
+    /// let kvs: Vec<usize> = (0..=5).collect();
+    ///
+    /// let _ = kvs
+    ///     .iter()
+    ///     .map(|k| amt.set(u64::try_from(*k).unwrap(), k.to_string()))
+    ///     .collect::<Vec<_>>();
+    ///
+    /// let kvs = kvs
+    ///     .iter()
+    ///     .map(|k| (k.clone(), k.to_string()))
+    ///     .collect::<Vec<_>>();
+    ///
+    /// // Read 2 elements.
+    /// let mut results = amt.iter().take(2).collect::<Result<Vec<_>, _>>().unwrap();
+    /// assert_eq!(results.len(), 2);
+    ///
+    /// // Read the rest.
+    /// if let Ok(mut res) = amt.iter_from(results.last().unwrap().0 + 1) {
+    ///     for res in &mut res {
+    ///         results.push(res.unwrap());
+    ///     }
+    /// }
+    ///
+    /// // Assert that we got out what we put in.
+    /// let results: Vec<_> = results
+    ///     .into_iter()
+    ///     .map(|(k, v)| (k.clone(), v.clone()))
+    ///     .collect();
+    ///
+    /// assert_eq!(kvs, results);
+    ///
+    /// # anyhow::Ok(())
+    /// ```
+    pub fn iter_from(&self, key: usize) -> Result<Iter<'_, V, &BS, Ver>, crate::Error> {
```

Comment:
We should be able to seek to the exact offset we want in logn time. We need to implement similar logic to `get` to skip entire sub-trees. I can also take a crack at implementing this if you don't feel comfortable with it.

Reply:
I can make another pass at it, but to simplify things, how about we address it in another PR since this one is quite large already?

Reply:
In that case, I'd remove it from this PR entirely. The implementation will likely be significantly different anyways.

Reply:
Sounds good, I'll open a tracking issue.

Reply:
#1923